GS-946 configurable reports enhancement

Sajonasj-OC opened the pull request:
## Description:
<!-- Describe your changes in detail / numbered list of distinct changes for reference -->
Issue: Configurable reports lacking event logs.
This PR will add 
- Reports now log when they’re viewed, exported, created, edited, duplicated, imported, or deleted.
- Report actions (view/export/import/duplicate/visibility) now record activity.
- Component edits and SQL changes trigger activity entries, and save failures now surface clearer errors.

## Checklist before requesting a review:
<!-- Remove non-applicable items e.g. epic sub-issue points -->
<!-- Add an 'x' inside the brackets to check the item -->

- [x] I have performed a self review of my code.
- [x] My code follows the [Moodle Coding Style](https://moodledev.io/general/development/policies/codingstyle).
- [x] Version numbers have been updated.
- [x] Code has been commented, particularly in hard-to-understand areas.

Sajonasj-OC committed:
GS-946: full event features working and early testing

Sajonasj-OC committed:
GS-946: cleaned up events: changed hardcoded strings to lang strings, added descriptions

Sajonasj-OC committed:
GS-946: isolated event codes to local helper file

Sajonasj-OC committed:
GS-946: added event language strings

Sajonasj-OC committed:
GS-946: added events to legacy files

Sajonasj-OC committed:
GS-946: added new classes for events, added new local helper file

Sajonasj-OC committed:
GS-946: minor styling changes

Sajonasj-OC committed:
GS-946: Fix configurable report event logging snapshots and rename helperlib to gchlolhelper

Sajonasj-OC committed:
GS-946: Polish report events and component error
- Add docblocks to the new report event classes.
- Remove backup/restore mappings from those events.
- Localize the component save error via  in editplugin.php.

Sajonasj-OC committed:
GS-946: Tidy configurable report events by adding docblocks and stdClass imports, removing backup/restore mappings and casts, and localizing the component save error.

Sajonasj-OC committed:
GS-946: updated comments

NickLambellGCH commented on gchlolhelper.php:
I think it would be nicer to see this as an autoloadable class with static functions rather than this older style include / global function format. It also allows you to have much cleaner function names as they don't need to be globally unique.
Following the standards, I think it would make the most sense for this class to go in `classes/local/util` and then probably be called `event_util` or something similar.

NickLambellGCH commented on gchlolhelper.php:
```diff
@@ -0,0 +1,421 @@
+ */
+function cr_log_report_viewed($context, stdClass $report, string $source = 'ui'): void {
+    if (
+        empty($report) ||
```
Since the `$report` param is typed as a `stdClass` it will never be `empty()` since even a completely blank object isn't considered empty.
Based off of a few of the use cases I looked at, it doesn't seem like `$report->id` would ever be not a valid value either so maybe this whole check could be removed.
This seems to apply to most of the event logging functions in this class.

It can be tempting to have a function validate its parameters in detail bit it can sometimes be excessive. Often times it makes the most sense to just have a function do what it is supposed to do in a 'perfect world' where it's params are correct and then just ensure it is implemented properly.
An exception to this is if you're writing 'api' kinds of functions, especially if designed to be used by external parties. But for internal code you can have a level of correct usage expectation :smile:

NickLambellGCH reviewed (changes requested):
Just a bit of structural feedback on this one, otherwise looking pretty good from the initial pass :smile:

Sajonasj-OC committed:
GS-946: Move report event helper into util class

- replace old helper with an autoloaded utility class
- update report pages and components to use it
- remove unnecessary guard checks

NickLambellGCH commented on classes/local/util/event_util.php:
```diff
@@ -0,0 +1,411 @@
+    /**
+     * Log report was viewed.
+     *
+     * @param context $context
```
Looks like you're missing an import for `context` which has PhpStorm complaining. The import should be `core\context` :smile:

NickLambellGCH commented on classes/local/util/event_util.php:
```diff
@@ -0,0 +1,411 @@
+                'source' => $source,
+            ],
+        ]);
+        $event->add_record_snapshot('block_configurable_reports', $report);
```
I was doing some quick testing and got this notice on my local site when creating a new report. Make sure to have _Display debug messages_ on in the developer debug admin options of your local site to get it too.
I'm thinking you could either just default the field's value to what the DB would default to _or_ just query to get the actual DB record which should always have all the field the snapshot expects.

```
Fields list in snapshot record does not match fields list in 'block_configurable_reports'. Record is missing fields: lastexecutiontime
line 862 of /lib/classes/event/base.php: call to debugging()
line 107 of /blocks/configurable_reports/classes/local/util/event_util.php: call to core\event\base->add_record_snapshot()
line 122 of /blocks/configurable_reports/classes/local/util/event_util.php: call to block_configurable_reports\local\util\event_util::log_report_created()
line 249 of /blocks/configurable_reports/editreport.php: call to block_configurable_reports\local\util\event_util::log_report_created_from_data()
```

NickLambellGCH reviewed (changes requested):
Just a couple of issues I found from testing and when looking at the code in PhpStorm by the looks of it, otherwise looking good 🚀

Sajonasj-OC committed:
GS-946: GS-946: Fix report_created snapshot mismatch in event util

Add missing core\context import for context references.

In log_report_created_from_data(), fetch the persisted block_configurable_reports row for event snapshots (includes lastexecutiontime), with a safe fallback default.

NickLambellGCH reviewed (dismissed):
Looks good to me. I'll just get @MattFeinQH to have a look over it as well 🚀

Sajonasj-OC commented:
@MattFeinQH just pinging for visibility

MattFeinQH commented on classes/event/report_created.php:
```diff
@@ -0,0 +1,70 @@
+ * @license     https://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+class report_created extends base {
+    protected function init(): void {
```
Missing PHPDoc for this method.

MattFeinQH commented on classes/event/report_deleted.php:
```diff
@@ -0,0 +1,76 @@
+ * @license     https://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+class report_deleted extends base {
+    protected function init(): void {
```
Also missing PHPDoc

MattFeinQH commented on classes/event/report_duplicated.php:
```diff
@@ -0,0 +1,72 @@
+ * @license     https://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+class report_duplicated extends base {
+    protected function init(): void {
```
Also missing PHPDoc

MattFeinQH commented on classes/event/report_exported.php:
```diff
@@ -0,0 +1,71 @@
+ * @license     https://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+class report_exported extends base {
+    protected function init(): void {
```
Also missing PHPDoc

MattFeinQH commented on classes/event/report_imported.php:
```diff
@@ -0,0 +1,71 @@
+ * @license     https://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+class report_imported extends base {
+    protected function init(): void {
```
Also missing PHPDoc

MattFeinQH commented on classes/event/report_updated.php:
```diff
@@ -0,0 +1,71 @@
+ * @license     https://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+class report_updated extends base {
+    protected function init(): void {
```
Also missing PHPDoc

MattFeinQH commented on classes/event/report_viewed.php:
```diff
@@ -0,0 +1,70 @@
+ * @license     https://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+class report_viewed extends base {
+    protected function init(): void {
```
Also missing PHPDoc

MattFeinQH commented on classes/local/util/event_util.php:
```diff
@@ -0,0 +1,422 @@
+     * Compare old and new SQL config and log when changed.
+     *
+     * @param stdClass $reportconfig
+     * @param stdClass $data
+     * @param array $components
+     * @return void
+     */
+    public static function log_sql_change_if_needed(stdClass $reportconfig, stdClass $data, array $components): void {
+        $oldsql = $components['customsql']['config']->querysql ?? '';
+        if (
+            isset($data->querysql) &&
+            $data->querysql !== $oldsql
+        ) {
+            self::log_sql_change($reportconfig);
+        }
+    }
+}
```
It doesn't look like this method is used. Can it be deleted?

MattFeinQH commented on classes/local/util/event_util.php:
I am thinking that the helper functions in this class (e.g. log_report_viewed) could be moved the actual event class. This is how Moodle seems to do it now as well, see https://github.com/moodle/moodle/blob/3fd359b05ab9a42a9185059efe3c46fab980bad0/public/reportbuilder/classes/event/audience_created.php#L56 for an example. The only difference I would do is change the return in phpdoc to \core\event\base as PHPstorm seems to complain :)

You could also have a single phpdoc for the class and remove the one for the file.

Also missing the "context" parameter type for $context on a bunch of the methods :)

MattFeinQH reviewed (changes requested):
Functionality looks good. Just a few style and restructuring suggestions :)

Sajonasj-OC committed:
GS-946: GS-946: Address PR review and minimise upstream footprint

- Move event-logging helpers from event_util into each event class as create_from_* factories (mirrors Moodle core pattern).
- Add init() docblocks, type-hint context params, drop unused log_sql_change_if_needed, collapse file/class docblock in event_util.
- In non-GCHLOL plugin files, drop top-of-file use imports, inline fully-qualified class names, and prefix added blocks with // GCHLOL GS-946.

Sajonasj-OC committed:
GS-946: GS-946: Clean up dev review nits on configurable report events

Sajonasj-OC committed:
GS-946: GS-946: Validate event data and drop defensive fallbacks

- Add validate_data() to each configurable report event to enforce required 'other' keys, matching Moodle core pattern.
- Drop ?? '' / ?? 0 / ?? 'unknown' fallbacks in get_description() now that validation guarantees presence.
- Remove superfluous MOODLE_INTERNAL checks from autoloaded event classes (linter flagged).
- Mark cr_import_xml return-value change in locallib.php with // GCHLOL GS-946 per non-GCHLOL plugin rule.

Sajonasj-OC committed:
GS-946: GS-946: Drop DB reads from event logging and canonicalise markers

- Change cr_import_xml to return stdClass|false so managereport.php can pass the inserted record straight to report_imported::create_from_report (removes create_from_id + its DB read).
- Simplify report_created::create_from_data to build the snapshot from the form data in memory instead of re-reading the DB.
- Reuse the $updatedreport returned by build_report_change_summary in the edit flow instead of re-fetching after update_record.
- Canonicalise all GS-946 markers to "// GCHLOL: GS-946." so they pass the Moodle inline-comment linter.

Sajonasj-OC committed:
GS-946: GS-946: Drop dead ?? '' guards on name in event factories

Sajonasj-OC committed:
GS-946: GS-946: GS-946: Simplify event factories and restore upstream lines

- Restore mid-file MOODLE_INTERNAL guard and trailing newline in editplugin.php (out-of-scope deletions).
- Drop unused $snapshot param from report_updated::create_from_report and its single caller.
- Collapse report_duplicated and report_created to one public factory each; callers now assign the insert id before triggering.
- Align @return docblocks with : self across all seven event classes.
- Inline prepare_report_duplicate at the editreport callsite.

Sajonasj-OC committed:
GS-946: GS-946: Trim event util guards, docblocks, and stale lang strings

- Drop clone/property_exists/$haschanges defensive patterns in
  event_util; callers check $change !== '' for no-change case.
- Remove event:changegeneric and errorsavingcomponent lang strings
  now that their code paths are gone.
- Switch event classes to `use context` (legacy alias) for
  consistency with surrounding context_course/context_system usage.
- Consolidate GCHLOL marker in editreport duplicate flow and
  canonicalise the viewreport export marker.

Sajonasj-OC committed:
GS-946: GS-946: Drop event snapshots and simplify tab-level change logging

- Remove add_record_snapshot() from all 7 report event classes to stop
  debug warnings when partial records are passed (e.g. lastexecutiontime
  missing from form data on create/import).
- Replace per-action component events (added/modified/reordered/deleted)
  with a single unconditional log_tab_change($report, $comp) per save.
  Permissions tab condition-expression form save is now logged too.

Sajonasj-OC committed:
GS-946: GS-946: Address post-review nits and drop unreachable URL-import event

- Update log_tab_change docblock now that the whitelist is gone.
- Collapse redundant $event temp in all 7 event factories.
- Drop speculative $source='ui' param (kept on report_imported).
- Guard permissions form save against no-op writes, matching SQL.
- Split assignment-in-condition in editreport duplicate flow.
- Drop URL-import event trigger; query param has no UI entry point,
  so the event can't be exercised or tested. File upload and
  repository import paths still log.

Sajonasj-OC commented:
I did a huge refactor along with the advised changes

GS-946 commits ebbc7cf → HEAD

Tested Manually each events. Here is the summary of changes.
I tried to minimise code changes


ebbc7cf Fix report_created snapshot mismatch — re-read inserted row from DB so add_record_snapshot() had all schema columns (e.g. lastexecutiontime) and stopped warning in debug.
7832eb3 PR review + footprint minimisation — moved helpers into per-event create_from_* factories (Moodle core pattern); in upstream files switched to inline FQNs with // GCHLOL GS-946 markers.
0ea17f6 Dev-review nits — trimmed over-engineered branches in report_deleted/report_duplicated, removed an unused customsql flag, added two missing lang strings.
5f8f405 Validate event data — added validate_data() to each event class; dropped ?? '' / ?? 0 fallbacks in get_description(); removed MOODLE_INTERNAL from autoloaded event classes.
3875f59 Drop DB reads from event logging — cr_import_xml() now returns stdClass|false; report_created builds from in-memory form data; edit flow reuses $updatedreport; canonicalised every marker to // GCHLOL: GS-946.
3f6e674 Drop ?? '' guards on $report->name — name is NOT NULL + validated; fallback was dead.
8cf7478 Simplify factories + restore upstream lines — restored a stray-deleted MOODLE_INTERNAL + newline in editplugin.php; collapsed report_duplicated/report_created to single public factories; aligned @return self docblocks; inlined a single-use helper.
932ea29 Trim guards + stale lang strings — removed clone/property_exists/$haschanges defensive code in event_util; deleted event:changegeneric and errorsavingcomponent strings; switched events to legacy use context alias for consistency.
a7d9676 Drop snapshots + simplify tab logging — removed add_record_snapshot() from all 7 event classes; replaced per-action component events with a single log_tab_change($report, $comp) per save; added Permissions main-form save hook.
48c606b Post-review nits + drop URL-import event — fixed log_tab_change docblock; collapsed $event temp in 7 factories; dropped speculative $source='ui'; added no-op guard to Permissions save; split assignment-in-condition in duplicate flow; removed the undocumented URL-import event trigger.
Overall direction: event-heavy + snapshot-validated + many helpers → minimal payloads, no DB reads, no snapshots, one coarse Report updated per tab, validate required fields only.

MattFeinQH commented on classes/event/report_created.php:
```diff
@@ -0,0 +1,107 @@
+     *
+     * @param context $context
+     * @param stdClass $report Report record (id required)
+     * @return self
```
Change "self" to "static".

MattFeinQH commented on classes/event/report_created.php:
```diff
@@ -0,0 +1,107 @@
+     * @param stdClass $report Report record (id required)
+     * @return self
+     */
+    public static function create_from_report(context $context, stdClass $report): self {
```
Change "self" to "base"

MattFeinQH commented on classes/event/report_created.php:
```diff
@@ -0,0 +1,107 @@
+     * @return self
+     */
+    public static function create_from_report(context $context, stdClass $report): self {
+        return self::create([
```
Change "self" to "static".

MattFeinQH reviewed (changes requested):
Nice work Jonas. The changes you made were good. Just some more pedantic changes after having a chat with Nick :)

We have decided it would be good moving forward to use "static" instead of "self" as it is generally preferred unless there is a specific reason why self must be used. So if you could change all instances of self to static that would be great.

I have also suggested some changes to the event helper functions to stop PHPStorm from complaining, as it thinks self::create is returning a base object instead of a self object. I have only added a comment for the first one as an example, however can you do the same for the rest of the event helper functions?

Sajonasj-OC committed:
GS-946: GS-946: Switch event factories to late static binding

- Use static:: instead of self:: for class constants and create() calls.
- Return type narrowed to base (matches core::create()), @return docblock
  set to static so IDEs infer the late-bound subclass.
